pkg/framework/capi-machinesets.go

```diff
@@ -0,0 +1,195 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type CAPIMachineSetParams struct {
+	msName            string
+	clusterName       string
+	failureDomain     string
+	replicas          int32
+	infrastructureRef corev1.ObjectReference
+}
+
+// NewCAPIcapiMachineSetParams returns a new CAPIMachineSetParams object.
+func NewCAPIMachineSetParams(msName, clusterName, failureDomain string, replicas int32, infrastructureRef corev1.ObjectReference) CAPIMachineSetParams {
+	Expect(msName).ToNot(BeEmpty())
+	Expect(clusterName).ToNot(BeEmpty())
+	Expect(infrastructureRef.APIVersion).ToNot(BeEmpty())
+	Expect(infrastructureRef.Kind).ToNot(BeEmpty())
+	Expect(infrastructureRef.Name).ToNot(BeEmpty())
+
+	return CAPIMachineSetParams{
+		msName:            msName,
+		clusterName:       clusterName,
+		replicas:          replicas,
+		infrastructureRef: infrastructureRef,
+		failureDomain:     failureDomain,
+	}
+}
+
+// CreateCAPIMachineSet creates a new MachineSet resource.
+func CreateCAPIMachineSet(ctx context.Context, cl client.Client, params CAPIMachineSetParams) (*clusterv1.MachineSet, error) {
+	By(fmt.Sprintf("Creating MachineSet %q", params.msName))
+
+	userDataSecret := "worker-user-data"
+
+	ms := &clusterv1.MachineSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "MachineSet",
+			APIVersion: "machine.openshift.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      params.msName,
+			Namespace: ClusterAPINamespace,
+		},
+		Spec: clusterv1.MachineSetSpec{
+			Replicas:    &params.replicas,
+			ClusterName: params.clusterName,
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"machine.openshift.io/cluster-api-cluster":    params.clusterName,
+					"machine.openshift.io/cluster-api-machineset": params.msName,
+				},
+			},
+			Template: clusterv1.MachineTemplateSpec{
+				ObjectMeta: clusterv1.ObjectMeta{
+					Labels: map[string]string{
+						"machine.openshift.io/cluster-api-cluster":    params.clusterName,
+						"machine.openshift.io/cluster-api-machineset": params.msName,
+					},
+				},
+				Spec: clusterv1.MachineSpec{
+					Bootstrap: clusterv1.Bootstrap{
+						DataSecretName: &userDataSecret,
+					},
+					ClusterName:       params.clusterName,
+					InfrastructureRef: params.infrastructureRef,
+				},
+			},
+		},
+	}
+
+	if params.failureDomain != "" {
+		ms.Spec.Template.Spec.FailureDomain = &params.failureDomain
+	}
```

Comment:
We could use the resourcebuilder for these https://github.com/openshift/cluster-api-actuator-pkg/tree/master/testutils/resourcebuilder